Updates to PDS4 branch

krlberry opened the pull request:
## Description
Adds: 
* isisimport, a generalized input application which currently supports pds4 input and a template to import tgo cassis images.
* Support for tgo casssis labels and the ability to look up the template to use for export to isisexport

Renames:
* topds4 to isisexport

## Related Issue
ISIS PDS4 work

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [ ] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [ ] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [ ] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

krlberry committed:
Initial isisimport application  (#4470)

* Initial isisimport application modeled to work on importing tgocassis images

* Add template to toisis

krlberry committed:
Draft cassis template file added to pds4 branch (#4472)

* Initial isisimport application modeled to work on importing tgocassis images

* Add template to toisis

* actual cassis label template

* Reverted accidently modified template

krlberry committed:
Add new locations of import and export templates and update cmake accordingly (#4477)

* Update cmake to new locations of import and export templates

* Fix mixed tabs and spaces

* Removed topds4 directory from CMakeLists

krlberry committed:
Updates XmlToJson parsing to be more inja-friendly (#4478)

* Updates XmlToJson parsing to indicate XML attributes and text values in a way friendly to inja parsing of the json

* Update XmlToJson tests

amystamile-usgs committed:
Tgocassis Template Completed for Instruments and Archive Groups (#4481)

* Updates to cassis label template

* Further updates to cassis template

* Updates to accomodate PR 4478 changes

* Cassis template completed for Instrument and Archive Groups.

krlberry committed:
Generalize PDS4 cube creation setup (#4485)

* Updates XmlToJson parsing to indicate XML attributes and text values in a way friendly to inja parsing of the json

* Update XmlToJson tests

* Generalized line/sample/band setup

tgiroux committed:
PDS4PixelType as a callback, TEMPLATE default option, and working TGOCaSSIS.tpl (#4479)

amystamile-usgs committed:
Cassis Template Updated BandBin and AlphaCube Groups (#4482)

* tgocassis Template Completed for AlphaCube and BandBin groups

* Additional template cleanup

* Updated Pixels Group

* Removed core object from template.

tgiroux committed:
protect template search and improve TGOCASSIS template (#4486)

Commit:
Renamed topds4 to isisexport (#4488)

amystamile-usgs committed:
Updates to Cassis Template and XML and Minor corrections to isisimport (#4490)

* Updates to cassis template. Corrections to isisimport to allow module tests to work properly.

* Added disclaimer to xml file.

kaitlyndlee committed:
Added mapping group template file (#4492)

* Added checks for existing keywords to be able to export an imported tgo image

* Added separate mapping group translation file

amystamile-usgs committed:
Callbacks for YearDOY and ConvertUniqueIdtoObservationId (#4491)

amystamile-usgs committed:
isisimport can now default to correct template using a file template. Updated Cassis template name. Minor updates in isisimport. (#4493)

amystamile-usgs committed:
Updates mapping group in template. Added callback to handle added Z to StartTime. (#4495)

tgiroux committed:
PDS4 - Fix template search and targetGroup template variable for tgocassis.tpl (#4494)

* fix template search and targetGroup template variable for tgocassis.tpl

* remove accidental extra tag

* address feedback from amy

amystamile-usgs committed:
Updated export template to produce parsable xml (#4497)

* Updated export template to be functioning xml to be used in isisimport.

* Added back 3D_Array but left in comments.

krlberry committed:
Adds tests with test data and fixes off-by-one error in isisimport (#4502)

* Added intial working section of tests

* Update test with test data and fix off-by-one in isisimport

* Add test input image

* Clean up from old design of test

amystamile-usgs commented on isis/src/base/apps/isisexport/isisexport.h:
```diff
@@ -0,0 +1,14 @@
+  extern void isisexport(Cube* cube, UserInterface &ui, Pvl *log=nullptr);
+  extern void isisexport(UserInterface &ui, Pvl *log=nullptr); }
+
+extern QString PDS4PixelType(Isis::PixelType ipixelType, Isis::ByteOrder ibyteOrder);
```
Should this be in the header? Just from doing app conversions, I noticed that the function prototypes are still in the apps cpp file rather than in the header. Not sure if we should be consistent with the other apps.

scsides replied:
We should be consistent

amystamile-usgs commented on isis/src/base/apps/isisimport/isisimport.xml:
```diff
@@ -0,0 +1,79 @@
+
+<application name="isisimport" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://isis.astrogeology.usgs.gov/Schemas/Application/application.xsd">
+  <brief>
+    Convert from PDS 4 format (other formats will be supported in the future) using template XML files to an ISIS Cube.
```
Should we standardize this and remove any mention of PDS 4 similar to changes in isisexport.xml?

scsides replied:
Yes

amystamile-usgs commented on isis/tests/FunctionalTestsIsisImport.cpp:
```diff
@@ -0,0 +1,226 @@
+
+static QString APP_XML = FileName("$ISISROOT/bin/xml/isisimport.xml").expanded();
+
+TEST_F(TempTestingFiles, FunctionalTestIsisImportLabelXmlInput) {
```
Not sure if we should keep the naming consistent with `FunctionalTestIsisexport`. So instead naming the tests `FunctionalTestIsisimport`. Same goes for the naming of the file.

amystamile-usgs reviewed (commented):
Added a few questions.

tgiroux commented on isis/tests/FunctionalTestsIsisexport.cpp:
```diff
@@ -71,7 +71,7 @@ TEST_F(SmallCube, FunctionalTestTopds4OriginalLabel) {
 }
 
-TEST_F(SmallCube, FunctionalTestTopds4NoOriginalLabel) {
+TEST_F(SmallCube, FunctionalTestIsisexport4NoOriginalLabel) {
```
Should be FunctionalTestIsisexportNoOriginalLabel (a 4 got left in)

scsides reviewed (changes requested):
We should make the suggested changes before merging.

jessemapel commented:
I looked at this again and I'm concerned about the topds4 name change. We put out topds4 in [5.0.0](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md#added) which means changing the name of the application is breaking. Is there someway we can provide a symlink in the bin directory for backwards compatibility? I don't think anyone used it yet, so maybe this is much ado about nothing.

krlberry committed:
Updated based on feedback

krlberry committed:
Moved local function into ISIS namespace

krlberry commented:
@jessemapel I'm not how to get a symlink to set up to `isisexport` from `topds4` in the build system. I agree that it's technically a breaking change, and also that it's unlikely that `topds4` is already being used. This is just a thought, but could it be possible to classify in the future (somehow) apps or even classes as "in beta" and  if they are "in beta" then we're allowed to make breaking changes to them without needing to increment the ISIS version number? It seems like one possible way to have the flexibility to distribute code that is still in development and might have major changes made to it, including API-breaking ones, but to get them available for (beta) testing in the community.

scsides commented:
I agree with the name change problem, but I also believe sooner is better than later is appropriate here. I don't know of a reasonable way to add a link.

Commit:
Ticked ISIS version to 6

Commit:
More tick to 6

jessemapel committed:
Updated CHANGELOG

We don't add the 6.0.0 heading until just before the actual release. I also added some missing periods.

jessemapel committed:
Merge branch 'dev' into pds4

jessemapel reviewed (dismissed):
6.0 tick looks good. I updated the CHANGELOG as we don't pull a new header until the release is frozen.

Commit:
Merge branch 'dev' into pds4

scsides reviewed (approved):
The most recent Jenkins build doesn't have any related failures.
